web/src/main/java/com/navercorp/pinpoint/web/service/AdminServiceImpl.java

```diff
@@ -71,26 +65,62 @@ public void removeInactiveAgents(int durationDays) {
         if (durationDays < MIN_DURATION_DAYS_FOR_INACTIVITY) {
             throw new IllegalArgumentException("duration may not be less than " + MIN_DURATION_DAYS_FOR_INACTIVITY + " days");
         }
-        Map<String, List<String>> inactiveAgentMap = new TreeMap<>(String::compareTo);
 
-        List<Application> applications = this.applicationIndexDao.selectAllApplicationNames();
-        Set<String> applicationNames = new TreeSet<>(String::compareTo);
-        // remove duplicates (same application name but different service type)
-        for (Application application : applications) {
-            applicationNames.add(application.getName());
+        List<String> applicationNames = this.applicationIndexDao.selectAllApplicationNames()
+                .stream()
+                .map(Application::getName)
+                .distinct()
+                .collect(Collectors.toList());
+        Collections.shuffle(applicationNames);
+
+        int index = 1;
+        for (String applicationName: applicationNames) {
+            logger.info("Cleaning {} ({}/{})", applicationName, index++, applicationNames.size());
+            removeInactiveAgentInApplication(applicationName, durationDays);
         }
-        for (String applicationName : applicationNames) {
-            List<String> agentIds = this.applicationIndexDao.selectAgentIds(applicationName);
-            Collections.sort(agentIds);
-            List<String> inactiveAgentIds = filterInactiveAgents(agentIds, durationDays);
-            if (CollectionUtils.hasLength(inactiveAgentIds)) {
-                inactiveAgentMap.put(applicationName, inactiveAgentIds);
+    }
+
+    @Override
+    public int removeInactiveAgentInApplication(String applicationName, int durationDays) {
+        int retry = 3;
+
+        while (retry-- > 0) {
+            try {
+                return removeInactiveAgentInApplication0(applicationName, durationDays);
+            } catch (Exception e) {
+                logger.error("Backoff to remove inactive agents in application {}", applicationName, e);
+                waitOneMinute();
             }
         }
-        // map may become big, but realistically won't cause OOM
-        // if it becomes an issue, consider deleting inside the loop above
-        logger.info("deleting {}", inactiveAgentMap);
-        this.applicationIndexDao.deleteAgentIds(inactiveAgentMap);
+        logger.error("Failed to remove inactive agents in application {}", applicationName);
+
+        return 0;
+    }
+
+    private void waitOneMinute() {
+        try {
+            Thread.sleep(60000);
+        } catch (Exception ignored) {}
+    }
+
+    private int removeInactiveAgentInApplication0(String applicationName, int durationDays) {
+        final BulkConsumer<String> remover = new BulkConsumer<>(agentIds -> {
```

Comment:
It seems that the code is difficult to understand because consumer has another consumer.